### PR TITLE
Add component helper

### PR DESCRIPTION
**Pull request by ry5n**
Adds a dust helper that addresses the issues I’ve run into working with UI components in Dust JS. In short, this allows:
- Declarative composition of component templates in a way similar to Web Components. This is accomplished by passing the body (or bodies) of the Dust helper as params to the template. Importantly, bodies are evaluated using the “outer” (view) context so we can treat their content as effectively part of the view.
- Data isolation: by default no data enters a component except through Dust params and bodies.
- Predictability: since the component context is isolated, Dust won’t walk the context stack looking for keys, preventing accidental key access and improving testability.

An example of the syntax this makes possible:

``` dust
{@component is="pinny" title=sidebarTitle}
    <p>Body content. This can be anything, including other components and no limit on nesting.</p>

    {! for example !}
    {#sidebarSections}
        {@component is="another-component" class="c--fancy"}
            {sectionContent}
        {/component}
    {/sidebarSections}
{:footer}
    <p>Components can have multiple bodies! (Additional bodies are named – in this case, `footer`.)</p>
{/component}
```

**Note**: This helper relies on component templates having been [compiled under a name](https://github.com/linkedin/dustjs/wiki/Dust-Tutorial#Compiling_a_Dust_template).

Status: **Opened for comment**

Reviewers: Tools team (@tedtate @donnielrt @jeffkamo) and any other interested folks 
Ticket: https://mobify.atlassian.net/browse/CSOPS-1277
Linked PRs: none
## Changes
- Adds component helper code; tests are already in place.
## How to test-drive this PR
1. Review code
2. Run tests
   - Clone this repo and `cd` in
   - `npm install`
   - `git checkout component-helper`
   - `grunt test`
## Research resources
- [Dust issues](https://gist.github.com/ry5n/1455a1dbf2b6ff56fcfd): short explainer document
- [Original attempt](https://github.com/mobify/dustjs/pull/2): this is the same idea, but is a fork of Dust itself (it does allow nicer syntax).
- [Ember Components](http://emberjs.com/api/classes/Ember.Component.html): a very similar idea
## Todo
- [ ] Engineer Review
- [ ] Engineer Approval

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -1 +1,102 @@
...
+    //   not passed to the template.
+    // - `body` is reserved as the key in which the default dust body is
+    //   provided to the template. It therefore can’t be used as a param name.
+    var reservedParamNames = ['name', 'ctx', 'body'];
+
+    // If and only if an explit context was provided, set up that context.
+    // Otherwise, the context available to the component template is limited
+    // to the params and bodies passed in.
+    if (params.ctx) {
+      componentContext.blocks = context.blocks;
```
Comment (ry5n) [inline index.js]: Just realised this should be params.ctx.blocks, but not sure if that will work.

File: index.js [hunk 2]
```diff
@@ -1 +1,102 @@
...
+
+      if (context.stack && context.stack.tail) {
+        // Grab the tail off of the previous context if we have it.
+        componentContext.stack = context.stack.tail;
+      }
+
+      // Reattach the head.
+      componentContext = componentContext.push(context.stack.head);
+    }
```
Comment (ry5n) [inline index.js:60]: Same with any reference to `context` in this if statement.

File: index.js [hunk 3]
```diff
@@ -1 +1,103 @@
...
+
+    bodies = bodies || {};
+    params = params || {};
+
+    // Keep a clear reference to the original context.
+    var originalContext = context;
+
+    // By default, components are completely isolated from the outside template
+    // context. So we start with a new, empty context.
+    var componentContext = dust.makeBase(context.global);
```
Comment (ry5n) [inline index.js:28]: This is copied from what Dust does internally for partials. Question is, is this global property always empty? If not, should it be used here?

File: index.js [hunk 4]
```diff
@@ -1 +1,103 @@
...
+    // Otherwise, the context available to the component template is limited
+    // to the params and bodies passed in.
+    if (params.ctx) {
+      componentContext = componentContext.push(params.ctx);
+
+      // The empty context created above is now set as the tail; we can safely
+      // remove it.
+      if (componentContext.stack) {
+        componentContext.stack.tail = undefined;
+      }
```
Comment (ry5n) [inline index.js:59]: If the componentContext created above can sometime have something in it (if context.global does) then we probably don’t want to do this.

**Comment by jeffkamo** [comment]
What is required to "complete" this? Do you have a set of requirements that must be fulfilled? Does it just need some engineer +1s?

**Comment by tedtate** [comment]
I would like to review this before we begin to integrate it into our projects.

**Comment by ry5n** [comment]
Yes, this needs a few things:
- Technical review. Despite myself I’ve been forging ahead; there is more up-to-date code in #2 now.
- Concept review. This is probably more important. I’m planning to give a presentation at an upcoming engineering meeting that will outline the problems this is solving, our current workflows on CST, and how this helps.

**Comment by ry5n** [comment]
@tedtate One thing that I recently realized (which I should have known) is that adaptive.js uses an older version (2.3) of Dust and shims its lack of AMD support. This work is based on Dust 2.6 which adds AMD support, so currently the two are incompatible (sad trombone).

Luckily that shouldn’t be hard to fix one way or another.

**Comment by jeffkamo** [comment]
Yikes. Is that going to completely derail this? Can we update new projects to use Dust 2.6?

**Comment by ry5n** [comment]
No no, it’s fine. If we don’t want to upgrade adaptive.js then I just need to wrangle dependencies and maybe change the UMD wrapper. The take-home is that there’s still plenty to do before we can start using this.

**Comment by jeffkamo** [comment]
Gotcha, thanks

**Comment by ry5n** [comment]
👋😛

**Pull request closed without merging**